src/pyflex/window_selector.py

```diff
@@ -318,32 +343,37 @@ def select_windows(self):
 
         self.calculate_preliminiaries()
 
+        self.determine_signal_and_noise_indices()
+        if self.config.check_global_data_quality:
+            if not self.check_data_quality():
+                return []
+
         # Perform all window selection steps.
         self.initial_window_selection()
         # Reject windows based on traveltime if event and station
         # information is given. This will also fill self.ttimes.
         if self.event and self.station:
-            self.reject_on_traveltimes()
+            self.reject_on_selection_mode()
         else:
             msg = "No rejection based on traveltime possible. Event and/or " \
                   "station information is not available."
             logger.warning(msg)
             warnings.warn(msg, PyflexWarning)
 
-        self.determine_signal_and_noise_indices()
-        if self.config.check_global_data_quality:
-            self.check_data_quality()
         self.reject_windows_based_on_minimum_length()
         self.reject_on_minima_water_level()
         self.reject_on_prominence_of_central_peak()
         self.reject_on_phase_separation()
+        self.__print_remaining_windows()
         self.curtail_length_of_windows()
+        self.__print_remaining_windows()
         self.remove_duplicates()
         # Call once again as curtailing might change the length of some
         # windows. Very cheap so can easily be called more than once.
         self.reject_windows_based_on_minimum_length()
         self.reject_based_on_signal_to_noise_ratio()
         self.reject_based_on_data_fit_criteria()
+        self.__print_remaining_windows()
```

Comment:
We'll eventually have to benchmark these calls. They all call `logger.debug` for each and every window. Sometimes the algorithm results in a ridiculous number of temporary windows where all these function calls might actually might make a difference. But that is just something to keep in mind for now.